### PR TITLE
Move exoplanet generation

### PR DESCRIPTION
**Pull request by PsyCommando**
Depends on: #2995 

## Description of changes
Splits exoplanet generation, processing, and data from their overmap marker.

* Moved exoplanet gen to a subclass of map_template, since they were a really good fit for abstracting level generation, and categorizing planets and etc..
* Fauna and flora gen/handling are now generated via separate datums. Those can be overridden and the overridden type can be specified in the map_template for allowing better control over how those work for specific planets. 
* Moved day cycle processing to a subsystem.
* Made weather processing handling mostly independent of exoplanets.  

**TODO**
* ~~Generalize planetoid data a bit more so it can be used downstream for planet loaded from a map file.~~
* Fix weather materials handling.
* ~~Fix planet rock colors.~~
* ~~Fix ambient lighting not applying properly.~~
* ~~Allow for lateral z-levels to generate and connect when specified.~~
* ~~Maybe move respawn handling to a "mob ai director" of sort, and make it independent of exoplanets. Maybe a subsystem?~~ (A bit out of scope for this PR.)
* ~~Major clean up and reorganization. I'm still mostly just at the prototype stage.~~
* ~~Add/tweak unit tests for exoplanets.~~ (Unit tests currently skip exoplanets.)
* ~~Fix strata conditions not being respected.~~

Also, I'm looking for some feedback, because some things I feel could be done a lot better, but I'm not quite sure what to focus on. 

## Why and what will this PR improve
With this change, exoplanets can be initialized at the same time as any other templates, since there's no dependencies on when the overmap initialize anymore.
And the overmap marker for exoplanet is now a lot less bloated with junk.

### CODE REVIEW

**Comment by PsyCommando** [comment]
@MistakeNot4892 uh, why did you mark it as needing a rebase?

**Comment by MistakeNot4892** [comment]
Because you had a `Level data datum + improv` commit that looked like a duplicate. I thought it needed rebasing since the dep got merged. :P

**Comment by MistakeNot4892** [comment]
Is this waiting on review or is it still WIP?

**Comment by PsyCommando** [comment]
> Is this waiting on review or is it still WIP?

Should be good for review!

File: code/controllers/subsystems/daycycle.dm
Comment (MistakeNot4892) [inline code/controllers/subsystems/daycycle.dm]: Weird leading linebreaks?

File: code/controllers/subsystems/weather.dm
Comment (MistakeNot4892) [inline code/controllers/subsystems/weather.dm]: Shouldn't this be in the loop below?

File: code/modules/xgm/xgm_gas_mixture.dm
Comment (MistakeNot4892) [inline code/modules/xgm/xgm_gas_mixture.dm:537]: Please cache this a la reagent mix colours, these RGB blend procs are costly.

File: code/modules/overmap/sectors.dm
Comment (MistakeNot4892) [inline code/modules/overmap/sectors.dm]: Is `map_z` always ordered? Are both of these lists always lists? This feels unsafe.

File: code/modules/overmap/planetoids/planetoid_skybox.dm
Comment (MistakeNot4892) [inline code/modules/overmap/planetoids/planetoid_skybox.dm]:
````
skybox_image.overlays = list(
    etc,
    etc,
    etc
)
````
would likely be a bit faster, though it's admittedly a microopt for this case.

File: code/modules/multiz/turf_mimic_edge.dm
Comment (MistakeNot4892) [inline code/modules/multiz/turf_mimic_edge.dm:204]: Heck of a proc name.

File: code/modules/multiz/turf.dm
Comment (MistakeNot4892) [inline code/modules/multiz/turf.dm]: What's with this file? Commented out entirely?

File: code/modules/multiz/level_data.dm
Comment (MistakeNot4892) [inline code/modules/multiz/level_data.dm]: Main purpose of this config option is to keep init times down for testing, which isn't really that important nowadays. I would prefer to see it generalized to disable all map generation (ie. before build_away_sites() and build_exoplanets()) rather than removed. Possibly rename it to `enable_roundstart_map_generation`.

File: code/modules/mining/mine_turfs.dm
Comment (MistakeNot4892) [inline code/modules/mining/mine_turfs.dm:52]: I have a branch somewhere to kill this turf type...

File: code/modules/materials/geology/_strata.dm
Comment (MistakeNot4892) [inline code/modules/materials/geology/_strata.dm]: This is the purpose of `base_materials` on `/decl/strata`.

**Comment by MistakeNot4892** [comment]
Github is struggling with this one. If you could split out stuff like the weather subsystem changes and the SSdaynight cycle addition into their own PRs it might chug less.

File: code/modules/xenoarcheaology/datums/engravings_flavor.dm
Comment (MistakeNot4892) [inline code/modules/xenoarcheaology/datums/engravings_flavor.dm:3]: From glancing over the code it looks like this could just be made a decl with a couple of changes from instantiating it to GET_DECL().

File: code/controllers/subsystems/weather.dm
Comment (MistakeNot4892) [inline code/controllers/subsystems/weather.dm]: Treating this like a lazylist, using lazyaccess etc. is potentially problematic since you're using numerical indices; I would just make them straight list access and assume it's a valid list, maybe reindex it on world z expansion like the other lists that use z level as a numerical index.

File: code/modules/multiz/turf.dm
Comment (PsyCommando) [reply code/modules/multiz/turf.dm]: I think something went wrong when I merged upstream. All that stuff got moved to code/modules/multiz/turf_mimic_edge.dm. I'll get rid of that comment block.

File: code/modules/multiz/turf_mimic_edge.dm
Comment (PsyCommando) [reply code/modules/multiz/turf_mimic_edge.dm:204]: It's really ``shared_mimic_edge_get_additional_vis_contents``, but it was getting a bit long. :P

File: code/modules/overmap/planetoids/planetoid_skybox.dm
Comment (PsyCommando) [reply code/modules/overmap/planetoids/planetoid_skybox.dm]: The main idea was just to split those things in smaller more manageable bits. 
And the skybox is generated exactly once per exoplanets. So idk how much it would matter yeah. 
Is that something you'd want changed?

File: code/modules/overmap/sectors.dm
Comment (PsyCommando) [reply code/modules/overmap/sectors.dm]: You're right about that. Currently the topmost level is usually the highest z in the sector's zlevels list because of how z stacks works. 

So, in short, easy way to fix this is to do ``max(map_z)`` instead of ``map_z[1]``. But, if we rework z stacks to have non-sequential up/down level connections eventually, that wouldn't work. 
It would be nice to have a standard way of accessing z stacks we could use instead of taking shortcuts like I did. Currently it's mostly done through ``global.z_levels`` and the ``/obj/abstract/map_data`` object, but those are kind of subject to change with z mimic ports.

File: code/modules/xgm/xgm_gas_mixture.dm
Comment (PsyCommando) [reply code/modules/xgm/xgm_gas_mixture.dm:537]: You mean ``/datum/reagents/proc/get_color()`` right?
I can do that, but /datum/gas_mixture are a lot less well encapsulated than /datum/reagents. A majority of things still change the gas list directly instead of using the procs, and knowing when to clear the cached color is a bit more unreliable.

File: code/controllers/subsystems/weather.dm
Comment (PsyCommando) [reply code/controllers/subsystems/weather.dm]: That's a good point..

File: code/controllers/subsystems/weather.dm
Comment (PsyCommando) [reply code/controllers/subsystems/weather.dm]: You mean adding a call in SSmapping's ``increment_world_z_size()`` proc and get SSweather to rebuild its weather_by_z list?
I could try that.

File: code/controllers/subsystems/daycycle.dm
Comment (PsyCommando) [reply code/controllers/subsystems/daycycle.dm]: It's probably just when I tried writing a description comment, and just ended up deleting it.

File: code/modules/materials/geology/_strata.dm
Comment (PsyCommando) [reply code/modules/materials/geology/_strata.dm]: I'll have to take a deeper look at this. I think there was a reason, but it's been a while. I'll see what I can do later today.

File: code/modules/mining/mine_turfs.dm
Comment (PsyCommando) [reply code/modules/mining/mine_turfs.dm:52]: By killing, do you mean you ported it's features to a more generic type of turf, or you just deleted it and all it's usages?

File: code/modules/multiz/level_data.dm
Comment (PsyCommando) [reply code/modules/multiz/level_data.dm]: When talking about map generation, do you mean solely like noise maps used to generate terrain features, or like map stitching randomly picked away sites/ruins? 
Because, I assumed the purpose of this var was to prevent random noise terrain gen from happening, seeing as it was preventing that on mining turfs?

File: code/modules/xenoarcheaology/datums/engravings_flavor.dm
Comment (PsyCommando) [reply code/modules/xenoarcheaology/datums/engravings_flavor.dm:3]: I could probably leave this as stub, and make another PR to decl-ify that?

File: code/modules/mining/mine_turfs.dm
Comment (MistakeNot4892) [reply code/modules/mining/mine_turfs.dm:52]: Unifying the features with /turf yeah.

File: code/modules/multiz/level_data.dm
Comment (MistakeNot4892) [reply code/modules/multiz/level_data.dm]: All gen of any kind that could bloat out init time. The sole purpose of the config var is to skip optional overhead.

File: code/controllers/subsystems/weather.dm
Comment (MistakeNot4892) [reply code/controllers/subsystems/weather.dm]: Yeah essentially.

File: code/modules/xenoarcheaology/datums/engravings_flavor.dm
Comment (MistakeNot4892) [reply code/modules/xenoarcheaology/datums/engravings_flavor.dm:3]: Sure.

File: code/modules/maps/template_types/random_exoplanet/random_planet_level_data.dm
Comment (out-of-phaze) [inline code/modules/maps/template_types/random_exoplanet/random_planet_level_data.dm:91]:
```suggestion
	if(!length(potentialRuins))
		log_world("Ruin loader was given no ruins to pick from.")
		return list()
	//#TODO: Fill in allowed area from a proc or something
```
This guards against `pick(candidates_ruins)` runtiming with `pick from empty list` if `potentialRuins` is empty.

**Review by out-of-phaze — CHANGES_REQUESTED** [review]
While testing I noticed some runtimes; these are the fixes to solve them.

atoms_init.dm - add this override to fix the runtimes with grass decals being deleted by ruin loading.
```dm
/atom/movable/EarlyDestroy(force = FALSE)
	loc = null // should NOT use forceMove, in order to avoid events
	return ..()
```

helper_landmarks.dm - move `/obj/abstract/landmark/clear/Initialize()` logic to `/obj/abstract/landmark/clear/LateInitialize()` to avoid a null reference error on `material.type` in `/turf/exterior/wall/proc/dismantle_wall()`.

reader.dm - add `/obj/abstract/landmark/exoplanet_spawn,` to `/dmm_suite/proc/types_to_delete()` in order to fix plants and animals spawning inside ruins. (This is used if `clear_contents = TRUE` and an object of the type exists on one of the turfs placed by the ruin.)

File: code/modules/maps/template_types/random_exoplanet/random_planet.dm
Comment (out-of-phaze) [inline code/modules/maps/template_types/random_exoplanet/random_planet.dm]:
```suggestion
/datum/map_template/planetoid/proc/build_z_stack(var/direction_from_root, var/list/adjacent_level_data, var/list/new_level_data_types, var/datum/planetoid_data/gen_data)
```

**Comment by PsyCommando** [comment]
> While testing I noticed some runtimes; these are the fixes to solve them.
> 
> atoms_init.dm - add this override to fix the runtimes with grass decals being deleted by ruin loading.
> 
> ```dm
> /atom/movable/EarlyDestroy(force = FALSE)
> 	loc = null // should NOT use forceMove, in order to avoid events
> 	return ..()
> ```
> 
> helper_landmarks.dm - move `/obj/abstract/landmark/clear/Initialize()` logic to `/obj/abstract/landmark/clear/LateInitialize()` to avoid a null reference error on `material.type` in `/turf/exterior/wall/proc/dismantle_wall()`.
>
> reader.dm - add `/obj/abstract/landmark/exoplanet_spawn,` to `/dmm_suite/proc/types_to_delete()` in order to fix plants and animals spawning inside ruins. (This is used if `clear_contents = TRUE` and an object of the type exists on one of the turfs placed by the ruin.)

Alright, thanks! I'll try to get around to fixing that this week hopefully.

**Comment by out-of-phaze** [comment]
I was wrong about the pick(candidates_ruins) runtime, it occurs after the last candidate is picked and removed. The check would need to be in the for loop or the for loop condition.

File: code/modules/multiz/level_data.dm
Comment (PsyCommando) [reply code/modules/multiz/level_data.dm]: But, won't that means we'll skip completely testing level gen during unit tests though? I'm seeing defines that force change the var to skip map gen as long as UNIT_TEST is defined.
Like, is it really worth the shortened testing time if we skip testing something kinda important like that? Picking and placing ruins, generating exoplanet terrain, generating mining data, etc..

EDIT: okay, I think I was wrong about that because the var is force changed after away sites and exoplanets are generated.

File: code/modules/materials/geology/_strata.dm
Comment (PsyCommando) [reply code/modules/materials/geology/_strata.dm]: Nah it's actually completely unused. So I'm removing it.

File: code/controllers/subsystems/weather.dm
Comment (PsyCommando) [reply code/controllers/subsystems/weather.dm]: Removed all the lazy stuff.

File: code/modules/overmap/planetoids/planetoid_skybox.dm
Comment (MistakeNot4892) [reply code/modules/overmap/planetoids/planetoid_skybox.dm]: No strong feelings, your call.

File: code/modules/overmap/planetoids/planetoid_skybox.dm
Comment (out-of-phaze) [inline code/modules/overmap/planetoids/planetoid_skybox.dm]: This runtimes, I tested this fix and it should work.
```suggestion
		var/datum/level_data/level_data = SSmapping.levels_by_z[lvl]
```

**Review by out-of-phaze — CHANGES_REQUESTED** [review]
Will do more testing with the updated version soon.

File: code/controllers/subsystems/daycycle.dm
Comment (PsyCommando) [reply code/controllers/subsystems/daycycle.dm]: fixed

File: code/controllers/subsystems/weather.dm
Comment (PsyCommando) [reply code/controllers/subsystems/weather.dm]: fixed

File: code/modules/xgm/xgm_gas_mixture.dm
Comment (PsyCommando) [reply code/modules/xgm/xgm_gas_mixture.dm:537]: done

File: code/modules/overmap/sectors.dm
Comment (PsyCommando) [reply code/modules/overmap/sectors.dm]: Fixed by trying to grab the map_data for the stack or the max zlevel index in map_z.

File: code/modules/overmap/planetoids/planetoid_skybox.dm
Comment (PsyCommando) [reply code/modules/overmap/planetoids/planetoid_skybox.dm]: I think I'll leave it like that for now.

File: code/modules/multiz/turf.dm
Comment (PsyCommando) [reply code/modules/multiz/turf.dm]: Removed the commented out code.

File: code/modules/multiz/level_data.dm
Comment (PsyCommando) [reply code/modules/multiz/level_data.dm]: Renamed the option and made it apply to all the obvious map gen stuff I could find.

File: code/modules/maps/template_types/random_exoplanet/random_planet.dm
Comment (PsyCommando) [reply code/modules/maps/template_types/random_exoplanet/random_planet.dm]: Woops, good catch XD

File: code/modules/overmap/planetoids/planetoid_skybox.dm
Comment (PsyCommando) [reply code/modules/overmap/planetoids/planetoid_skybox.dm]: Kk

File: code/modules/maps/template_types/random_exoplanet/random_planet.dm
Comment (PsyCommando) [reply code/modules/maps/template_types/random_exoplanet/random_planet.dm]: Applied it manually since I had to rename it in a bunch of places.

**Comment by PsyCommando** [comment]
> Github is struggling with this one. If you could split out stuff like the weather subsystem changes and the SSdaynight cycle addition into their own PRs it might chug less.

Hmm, well, I've looked at it, and it doesn't seem like the amount of work needed to port that to the old system really makes sense here? If I make it a stand alone PR, I have to make it compile and test it works, which is at least a 1-2 weeks job. And then I have to re-implement it and fix conflicts so it works with the new changes and test again.

Is there any ways we could reach a compromise on this?

File: code/modules/maps/template_types/random_exoplanet/fauna_generator.dm
Comment (comma) [inline code/modules/maps/template_types/random_exoplanet/fauna_generator.dm:224]: Wouldn't this cause threshold to drop p fast even if not a lot of animals are killed? If just 1 animal dies per 5 minutes it'll be depopulated in 20 minutes (right away in casae of megafauna)

File: code/modules/maps/template_types/random_exoplanet/random_planet_subtemplates.dm
Comment (comma) [inline code/modules/maps/template_types/random_exoplanet/random_planet_subtemplates.dm]: Maybe check that we do have other zlevels here before skipping? Otherwise there's 25% chance of a planet without any ruins placed

File: code/modules/maps/template_types/random_exoplanet/fauna_generator.dm
Comment (PsyCommando) [reply code/modules/maps/template_types/random_exoplanet/fauna_generator.dm:224]: I don't really know, that was the old fauna code I ported into the fauna generator. So I assumed it worked.

File: code/modules/maps/template_types/random_exoplanet/random_planet_subtemplates.dm
Comment (PsyCommando) [reply code/modules/maps/template_types/random_exoplanet/random_planet_subtemplates.dm]: That's a fair point. But I don't really know what's the expectation on exoplanet ruins. Like what's the minimum?

Comment (comma) [reply code/modules/maps/template_types/random_exoplanet/random_planet_subtemplates.dm]: Non zero at least, I don't see the need for this check at all tbh. Empty planets are boring to visit, we got budget to limit  amount of features, no need for random chance to skip budgeting completely

File: code/modules/maps/template_types/random_exoplanet/fauna_generator.dm
Comment (comma) [reply code/modules/maps/template_types/random_exoplanet/fauna_generator.dm:224]: Well, if that's old code can be figured out later I guess

File: code/modules/maps/template_types/random_exoplanet/random_planet_subtemplates.dm
Comment (PsyCommando) [reply code/modules/maps/template_types/random_exoplanet/random_planet_subtemplates.dm]: The check is so ruins aren't placed only on the surface level.

Comment (comma) [reply code/modules/maps/template_types/random_exoplanet/random_planet_subtemplates.dm]: It doesn't check if it's on surface or not, there's still 25% to skip the lower zlevel too, or skip entirely if it's 1z planet
There's some sort of budget splitting right after the skip too, does that not make it so ruins aren't placed only on surface?

Also pretty much all ruins are mapped with being surface in mind, I think their distribution should be addressed later with some sort of tag setup, don't need to force them to be underground this hard

Comment (out-of-phaze) [reply code/modules/maps/template_types/random_exoplanet/random_planet_subtemplates.dm]: Every single ruin except the drill site and *maybe*  the radioactive anomaly shrine and the monolith expects to be placed on the surface, having either been built there or crashed on the surface.

Comment (PsyCommando) [reply code/modules/maps/template_types/random_exoplanet/random_planet_subtemplates.dm]: Yeah, I changed it a few hours ago to just put all ruins onto the surface for now.

**Review by comma — DISMISSED** [review]

**Review by comma — APPROVED** [review]

**Review by out-of-phaze — APPROVED** [review]
Will merge and PR a fix for chlorine exoplanet sprites later, since it's not gamebreaking.

**Pull request merged**
